### PR TITLE
Port to CategoricalArrays 0.8

### PR DESCRIPTION
**Pull request by nalimilan**
Replace `CategoricalString` with `CategoricalValue{String}` and stop using `CategoricalArrays.index`. This simplifies the code a bit. Also drop some specializations which are equivalent to optimized methods defined in CategoricalArrays.

### CODE REVIEW

File: src/abstractdataframe/abstractdataframe.jl [hunk 1]
```diff
@@ -719,16 +719,13 @@ function completecases(df::AbstractDataFrame, col::Colon=:)
     end
     res = trues(size(df, 1))
     for i in 1:size(df, 2)
-        _nonmissing!(res, df[!, i])
+        res .&= .!ismissing.(df[!, i])
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:722]: you can remove `_nonmissing` definition above.

File: src/abstractdataframe/reshape.jl [hunk 2]
```diff
@@ -209,15 +206,15 @@ function _unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int,
             end
             i = length(unstacked_val[1])
         else
-            i = refkeycol_order[refkref]
+            i = refkref
```
Comment (bkamins) [inline src/abstractdataframe/reshape.jl:209]: maybe add type restrictions for `_unstack` parameters?

File: src/dataframerow/dataframerow.jl [hunk 3]
```diff
@@ -278,26 +278,13 @@ Base.merge(a::DataFrameRow, b::DataFrameRow) = merge(NamedTuple(a), NamedTuple(b
...
-        hash(CategoricalArrays.index(v.pool)[ref], h)
-    end
-end
-
 # hash of DataFrame rows based on its values
 # so that duplicate rows would have the same hash
 # table columns are passed as a tuple of vectors to ensure type specialization
 rowhash(cols::Tuple{AbstractVector}, r::Int, h::UInt = zero(UInt))::UInt =
-    hash_colel(cols[1], r, h)
+    hash(cols[1][r], h)
```
Comment (bkamins) [inline src/dataframerow/dataframerow.jl:285]: I understand that now it is not needed and proper hashing is ensured by CategoricalArrays.jl?

File: src/deprecated.jl [hunk 4]
```diff
@@ -18,8 +18,8 @@ function DataFrame(column_eltypes::AbstractVector{T}, cnames::AbstractVector{Sym
     end
     for i in eachindex(categorical)
         categorical[i] || continue
-        elty = CategoricalArrays.catvaluetype(nonmissingtype(updated_types[i]),
-                                              CategoricalArrays.DefaultRefType)
+        elty = CategoricalValue{nonmissingtype(updated_types[i]),
```
Comment (bkamins) [inline src/deprecated.jl:21]: this deprecation is 1 year old - maybe just remove this method from deprecated.jl?

File: test/broadcasting.jl [hunk 5]
```diff
@@ -699,9 +699,9 @@ end
         @test levels(df.c1) == levels(v)
         @test levels(df.c1) !== levels(v)
         df[!, :c2] .= v[2]
-        @test df.c2 == get.([v[2], v[2], v[2]])
+        @test df.c2 == fill(v[2], 3)
         @test df.c2 isa CategoricalVector
-        @test levels(df.c2) != levels(v)
+        @test levels(df.c2) == levels(v)
```
Comment (bkamins) [inline test/broadcasting.jl:704]: I understand this is a consequence of copying of all levels by `setindex!`?

**Comment by bkamins** [comment]
Looks good apart from what I commented on. Thank you!

File: src/deprecated.jl (hunk 4 above)
Comment (nalimilan) [reply src/deprecated.jl:21]: You know I love keeping old deprecated methods. :-) I'd say let's remove them all at once later if we want.

File: test/broadcasting.jl (hunk 5 above)
Comment (nalimilan) [reply test/broadcasting.jl:704]: Yes, that's a more logical behavior than before.

File: src/dataframerow/dataframerow.jl (hunk 3 above)
Comment (nalimilan) [reply src/dataframerow/dataframerow.jl:285]: Yes. Actually I'm not even sure it was required before. It's slightly faster, but not that much, and working with `DataFrameRow` is inefficient anyway (we have much more efficient code when hashing a full column, based on hashing each level only once). What do you think?

```julia
julia> df = categorical(DataFrame(rand(["a", "b", "c"], 10, 10)));

julia> r = view(df, 1, :);

# master
julia> @btime hash($r);
  471.533 ns (7 allocations: 352 bytes)

# this PR
julia> @btime hash($r);
  552.241 ns (7 allocations: 352 bytes)

Comment (bkamins) [reply src/dataframerow/dataframerow.jl:285]: It is OK given it is hardly used anyway.

**Pull request merged**